cranelift/codegen/src/dominator_tree.rs

```diff
@@ -242,88 +290,118 @@ impl DominatorTree {
+        loop {
+            match self.dfs_worklist.pop() {
+                Some(TraversalEvent::Enter(parent, block)) => {
+                    let node = &mut self.nodes[block];
+                    // The node has been visited
+                    if node.pre_number != 0 {
+                        continue;
+                    }
+
+                    self.dfs_worklist.push(TraversalEvent::Exit(block));
+
+                    let pre_number = self.stree.push(parent, block);
+                    node.pre_number = pre_number;
+
+                    // Use the same traversal heuristics as in traversals.rs
+                    self.dfs_worklist.extend(
+                        func.block_successors(block)
+                            .rev()
+                            .filter(|successor| self.nodes[*successor].pre_number == 0)
+                            .map(|successor| TraversalEvent::Enter(pre_number, successor)),
+                    );
+                }
+                Some(TraversalEvent::Exit(block)) => self.postorder.push(block),
+                None => break,
+            }
+        }
     }
 
-    /// Build a dominator tree from a control flow graph using Keith D. Cooper's
-    /// "Simple, Fast Dominator Algorithm."
-    fn compute_domtree(&mut self, func: &Function, cfg: &ControlFlowGraph) {
-        // During this algorithm, `rpo_number` has the following values:
-        //
-        // 0: block is not reachable.
-        // 1: block is reachable, but has not yet been visited during the first pass. This is set by
-        // `compute_postorder`.
-        // 2+: block is reachable and has an assigned RPO number.
-
-        // We'll be iterating over a reverse post-order of the CFG, skipping the entry block.
-        let (entry_block, postorder) = match self.postorder.as_slice().split_last() {
-            Some((&eb, rest)) => (eb, rest),
-            None => return,
-        };
-        debug_assert_eq!(Some(entry_block), func.layout.entry_block());
-
-        // Do a first pass where we assign RPO numbers to all reachable nodes.
-        self.nodes[entry_block].rpo_number = 2 * STRIDE;
-        for (rpo_idx, &block) in postorder.iter().rev().enumerate() {
-            // Update the current node and give it an RPO number.
-            // The entry block got 2, the rest start at 3 by multiples of STRIDE to leave
-            // room for future dominator tree modifications.
-            //
-            // Since `compute_idom` will only look at nodes with an assigned RPO number, the
-            // function will never see an uninitialized predecessor.
-            //
-            // Due to the nature of the post-order traversal, every node we visit will have at
-            // least one predecessor that has previously been visited during this RPO.
-            self.nodes[block] = DomNode {
-                idom: self.compute_idom(block, cfg).into(),
-                rpo_number: (rpo_idx as u32 + 3) * STRIDE,
+    /// Eval-compress
```

Comment:
This name ("eval-compress") doesn't mean anything to the reader -- could you give a more descriptive doc-comment?